### PR TITLE
devutils reciter + audio play

### PR DESCRIPTION
**Pull request by muhajirdev**
### Summary
 devutil reciter + audioplay

### CODE REVIEW

**Comment by mmahalwy** [comment]
#### 📝Changed routes:


| [**/\[chapterId\]**](https://quran-com-frontend-next-cf47oybft-qurancom.vercel.app/[chapterId]) |
|:-:|
|<a href="https://quran-com-frontend-next-cf47oybft-qurancom.vercel.app/[chapterId]">Screenshot of /[chapterId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/2JjkMVJRK25Fy79h7xQr99mImdr-acf6lQiXBojQ2yek-dzXipmimkDYxP0qXMOzJ3Phzmsa8KL_cqFZuNDJzg.png">(view full size)</a> |







Commit aabab5cdd0f7e1c6e2164b3e17188399742b5cca (https://quran-com-frontend-next-cf47oybft-qurancom.vercel.app).

File: src/api.ts [hunk 1]
```diff
@@ -89,6 +95,34 @@ export const getAvailableLanguages = async (language: string): Promise<Languages
   return camelizeKeys(payload) as LanguagesResponse;
 };
 
+/**
+ * Get the current available reciters with the name translated in the current language.
```
Comment (tifa2UP) [inline src/api.ts:99]: nit: no support for languages yet

File: src/api.ts [hunk 2]
```diff
@@ -89,6 +95,34 @@ export const getAvailableLanguages = async (language: string): Promise<Languages
...
+ *
+ * @returns {Promise<RecitersResponse>}
+ */
+export const getAvailableReciters = async (): Promise<RecitersResponse> => {
+  const payload = await fetcher(makeRecitersUrl());
+
+  return camelizeKeys(payload) as RecitersResponse;
+};
+
+export const getReciterAudio = async (
```
Comment (tifa2UP) [inline src/api.ts:109]: getAudioUrl

File: src/api.ts [hunk 3]
```diff
@@ -4,6 +4,9 @@ import {
   SearchResponse,
   AdvancedCopyRawResultResponse,
   LanguagesResponse,
+  RecitersResponse,
+  ReciterAudioResponse,
```
[the remark this answers is not in the document]
Comment (tifa2UP) [reply src/api.ts:8]: rename to: AudioUrl response

File: src/api.ts [hunk 4]
```diff
@@ -89,6 +95,34 @@ export const getAvailableLanguages = async (language: string): Promise<Languages
...
+export const getReciterAudio = async (
+  reciterId: number,
+  chapter,
+): Promise<ReciterAudioResponse> => {
+  const payload = await fetcher(makeReciterAudioUrl(reciterId, chapter));
+
+  return camelizeKeys(payload) as ReciterAudioResponse;
+};
+
+export const getVerseAudioTimestamp = async (
```
Comment (tifa2UP) [inline src/api.ts:118]: Add a comment explaining the function. 

Rename to: getVerseTimestamps

File: src/components/AudioPlayer/AudioPlayer.tsx [hunk 5]
```diff
@@ -28,6 +32,7 @@ const AudioPlayer = () => {
   const isMinimized = visibility === AudioPlayerVisibility.Minimized;
   const isExpanded = visibility === AudioPlayerVisibility.Expanded;
   const audioPlayerEl = useRef(null);
+  const audioData = useAudioData();
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioPlayer.tsx:35]: come back to this.

File: src/components/AudioPlayer/AudioPlayer.tsx [hunk 6]
```diff
@@ -58,8 +63,10 @@ const AudioPlayer = () => {
         currentRef.removeEventListener('pause', onAudioPause);
         currentRef.removeEventListener('ended', onAudioEnded);
       }
+      // reset isPlaying state
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioPlayer.tsx:66]: come back to this

File: src/components/AudioPlayer/AudioPlayer.tsx [hunk 7]
```diff
@@ -103,6 +110,15 @@ const AudioPlayer = () => {
...
   );
 
+  // TODO: Tempoary hack, need to implement proper solution. Maybe something like this https://github.com/E-Kuerschner/useAudioPlayer
+  if (typeof window !== 'undefined') {
+    // @ts-ignore
+    window.setTime = (time) => {
+      setTime(time);
+      play();
+    };
+  }
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioPlayer.tsx:120]: This should happen through redux.

File: src/components/AudioPlayer/AudioPlayer.tsx [hunk 8]
```diff
@@ -120,7 +136,7 @@ const AudioPlayer = () => {
       <div className={styles.innerContainer}>
         {/* We have to create an inline audio player and hide it due to limitations of how safari requires a play action to trigger: https://stackoverflow.com/questions/31776548/why-cant-javascript-play-audio-files-on-iphone-safari */}
         <audio
-          src="https://server12.mp3quran.net/tnjy/004.mp3"
+          src={audioData.url}
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioPlayer.tsx:139]: should come from redux.

File: src/pages/[chapterId]/index.tsx [hunk 9]
```diff
@@ -16,9 +18,16 @@ type ChapterProps = {
 };
 
 const Chapter: NextPage<ChapterProps> = ({ chapterResponse, versesResponse, hasError }) => {
+  const dispatch = useDispatch();
+  useEffect(() => {
+    dispatch(setChapter(Number(chapterResponse.chapter.id)));
+  }, [dispatch, chapterResponse.chapter?.id]);
+
```
Comment (tifa2UP) [inline src/pages/[chapterId]/index.tsx:25]: we're going to remove this.

File: src/redux/slices/AudioPlayer/state.ts [hunk 10]
```diff
@@ -1,11 +1,35 @@
...
+export type Audio = {
+  url: string;
+  totalDuration: number;
+};
+
 export type AudioState = {
   isPlaying: boolean;
   currentTime: number;
+  reciter?: Reciter;
+  audio: Audio;
```
Comment (tifa2UP) [inline src/redux/slices/AudioPlayer/state.ts:19]: change to AudioFile to match the API.

File: src/redux/slices/AudioPlayer/state.ts [hunk 11]
```diff
@@ -1,11 +1,35 @@
 import { createSlice, PayloadAction } from '@reduxjs/toolkit';
 
+// TODO: The filesize is getting big. need to split into another file
+
+type Reciter = {
```
Comment (tifa2UP) [inline src/redux/slices/AudioPlayer/state.ts:5]: this should go to the API definition.

File: src/redux/slices/AudioPlayer/state.ts [hunk 12]
```diff
@@ -1,11 +1,35 @@
...
+  audio: Audio;
+  chapter: number;
 };
 
-const initialState: AudioState = { isPlaying: false, currentTime: 0 };
+const initialState: AudioState = {
+  isPlaying: false,
+  currentTime: 0,
+  reciter: null,
+  chapter: 1,
```
Comment (tifa2UP) [inline src/redux/slices/AudioPlayer/state.ts:27]: remove

File: src/redux/slices/AudioPlayer/state.ts [hunk 13]
```diff
@@ -17,6 +41,21 @@ export const audioPlayerStateSlice = createSlice({
...
     },
+    setReciter: (state, action: PayloadAction<Reciter>) => {
+      return {
+        ...state,
+        reciter: action.payload,
+      };
+    },
+    setChapter: (state, action: PayloadAction<number>) => {
+      return { ...state, chapter: action.payload };
+    },
```
Comment (tifa2UP) [inline src/redux/slices/AudioPlayer/state.ts:52]: remove

File: types/Reciter.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+interface Reciter {
+  id: number;
+  name: string;
+  recitationStyle: string;
+  relativePath: string;
+}
```
Comment (tifa2UP) [inline types/Reciter.ts:6]: use this inside of the redux slice

File: src/components/AudioPlayer/AudioPlayer.tsx [hunk 15]
```diff
@@ -19,6 +22,7 @@ import Slider from './Slider';
 // import AudioKeyBoardListeners from './AudioKeyboardListeners';
 import MediaSessionApiListeners from './MediaSessionAPIListeners';
 import styles from './AudioPlayer.module.scss';
+import useAudioData from './useAudioData';
```
Comment (tifa2UP) [inline src/components/AudioPlayer/AudioPlayer.tsx:25]: this functionality should not be living here. The components that plays the audio should call the api, extract the url, and then pass than url onto redux.

File: src/components/QuranReader/PlayButton.tsx [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import { useDispatch } from 'react-redux';
+import { setAudioFile } from 'src/redux/slices/AudioPlayer/state';
+import styles from './PlayButton.module.scss';
+import Button from '../dls/Button/Button';
+
+interface Props {
+  chapterId: number;
+}
+const PlayButton = (props: Props) => {
```
Comment (tifa2UP) [inline src/components/QuranReader/PlayButton.tsx:9]: nit: PlayChapterAudioButton

File: src/components/Verse/VerseActions.tsx [hunk 17]
```diff
@@ -29,6 +34,12 @@ const VerseActions: React.FC<Props> = ({ verse }) => {
         verse={verse}
         setActiveVerseActionModal={setActiveVerseActionModal}
       />
+      <Button
```
Comment (tifa2UP) [inline src/components/Verse/VerseActions.tsx:37]: create a component called PlayVerseAudioButton

File: src/redux/slices/AudioPlayer/state.ts [hunk 18]
```diff
@@ -1,11 +1,57 @@
...
+const initialState: AudioState = {
+  isPlaying: false,
+  currentTime: 0,
+  audio: null,
+  reciter: {
+    id: 5,
+    name: 'Mishari Rashid al-`Afasy',
+    recitationStyle: 'Warsh',
+    relativePath: 'mishaari_raashid_al_3afaasee',
+  },
```
Comment (tifa2UP) [inline src/redux/slices/AudioPlayer/state.ts:22]: Move this to a constant on top of the file. 

```
const DEFAULT_RECITER = {
    id: 5,
    name: 'Mishari Rashid al-`Afasy',
    recitationStyle: 'Warsh',
    relativePath: 'mishaari_raashid_al_3afaasee',
  }
```

File: src/redux/slices/AudioPlayer/state.ts [hunk 19]
```diff
@@ -1,11 +1,57 @@
...
+};
+
+export const selectAudioPlayerState = (state) => state.audioPlayerState;
+export const selectReciter = (state) => state.audioPlayerState.reciter;
+export const selectAudioUrl = (state) => state.audioPlayerState.audio?.audioUrl;
+
+export const setAudioFile = createAsyncThunk<AudioFile, number>(
```
Comment (tifa2UP) [inline src/redux/slices/AudioPlayer/state.ts:29]: Add a small comment explaining what the function does.

**Review by tifa2UP — APPROVED** [review]

**Pull request merged**